### ISSUE TITLE
Issue on training tiny yolo voc

### ISSUE THREAD

**Issue by EnricoBeltramo** [AlexeyAB/darknet#268]
I make a training using tiny-yolo-voc.cfg customized  for 4 classes (car,dog,cat,person), but after 1000 iteration (last log: 1021: 0.637612, 0.947500 avg, 0.001000 rate, 10.991000 seconds, 65344 images), I still haven't any detection.
to train I used follow command:
darknet.exe detector train data/voc_custom.data tiny-yolo-voc_custom.cfg darknet19_448.conv.23

To detect, I tried both commands:
darknet.exe detector test data/voc_custom.data tiny-yolo-voc_custom.cfg backup/tiny-yolo-voc_custom_1000.weights dog.jpg -i 0
darknet.exe detector test data/voc_custom.data tiny-yolo-voc_custom.cfg backup/tiny-yolo-voc_custom_1000.weights dog.jpg -tresh 0.01

I attach here the cfg file used:
[net]
batch=64
subdivisions=8
width=416
height=416
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
max_batches = 40200
policy=steps
steps=-1,100,20000,30000
scales=.1,10,.1,.1

[convolutional]
batch_normalize=1
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=256
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=1
filters=512
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=1

[convolutional]
batch_normalize=1
filters=1024
size=3
stride=1
pad=1
activation=leaky

###########

[convolutional]
batch_normalize=1
size=3
stride=1
pad=1
filters=1024
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=45
activation=linear

[region]
anchors = 1.08,1.19,  3.42,4.41,  6.63,11.38,  9.42,5.11,  16.62,10.52
bias_match=1
classes=4
coords=4
num=5
softmax=1
jitter=.2
rescore=1

object_scale=5
noobject_scale=1
class_scale=1
coord_scale=1

absolute=1
thresh = .6
random=1

and voc_custom.data:
classes= 4
train  = data/voc/train.txt
valid  = data/voc/2007_test.txt
names = data/voc_custom.names
backup = backup/

and voc_custom.names:
car
cat
dog
person

Doing same thing with yolo-voc.2.0.cfg, it works (a good detection after 2000 steps). It works using standard weigths yolo-voc.weights and tiny-yolo-voc.weights: not only with my custom trained tiny-yolo-voc.
I use Win10, Visual studio 2015, CUDA 8, openCV 3.2, cnn
What should be the error?
Thank you!

**Comment by shizukachan** [comment]
you are initializing the Darknet Reference Network with Darknet19_448 weights, so your classifier network is essentially starting with random weights. so of course, the best course of action is to train more ;-)

Actually not. Use darknet partial to split off the conv layers of a pre-trained Darknet (see http://pjreddie.com/media/files/darknet.weights) instead.

darknet.conv.13 should be 25193940 bytes long.

A hint is to consider "burning in" the learning rate of the initial 1000 iterations by replacing the tiny-yolo-voc training configuration with this one:
learning_rate=0.001
burn_in=1000
max_batches = 40200
policy=steps
steps=20000,30000
scales=.1,.1

This should hopefully lead to slightly more stable training when starting from pre-trained weights.

**Comment by EnricoBeltramo** [comment]
@shizukachan Thank you for your suggestion! I tried both solution (to start from correct pretrained weights and to use burn_in option in firsth 1000 steps). 
I'm still training, but unfortunately not improvement about detection.
What I observe is that the avg decreased well, Avg Recall is high but Obj and No Obj are always 0.
I post a partial log of the training:

Region Avg IOU: 0.654453, Class: 0.909443, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.833333,  count: 18
Region Avg IOU: 0.609420, Class: 0.921004, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.750000,  count: 16
Region Avg IOU: 0.552798, Class: 0.966616, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 1.000000,  count: 2
Region Avg IOU: 0.538751, Class: 0.782769, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.625000,  count: 8
Region Avg IOU: 0.521379, Class: 0.985704, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.454545,  count: 11
Region Avg IOU: 0.644242, Class: 0.887315, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.933333,  count: 15
Region Avg IOU: 0.671180, Class: 0.749017, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 1.000000,  count: 6
Region Avg IOU: 0.770410, Class: 0.940213, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 1.000000,  count: 9
1561: 0.367512, 0.356705 avg, 0.001000 rate, 8.550000 seconds, 99904 images
Loaded: 0.000000 seconds
Region Avg IOU: 0.695436, Class: 0.952719, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 1.000000,  count: 7
Region Avg IOU: 0.613909, Class: 0.994981, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.875000,  count: 8
Region Avg IOU: 0.710797, Class: 0.919221, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 1.000000,  count: 8
Region Avg IOU: 0.570166, Class: 0.816978, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 1.000000,  count: 3
Region Avg IOU: 0.600714, Class: 0.932409, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.764706,  count: 17
Region Avg IOU: 0.647881, Class: 0.718039, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.750000,  count: 8
Region Avg IOU: 0.733311, Class: 0.799736, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 1.000000,  count: 4
Region Avg IOU: 0.605391, Class: 0.962357, Obj: 0.000000, No Obj: 0.000000, Avg Recall: 0.833333,  count: 12
1562: 0.221125, 0.343147 avg, 0.001000 rate, 9.250000 seconds, 99968 images

**Comment by shizukachan** [comment]
This is the sort of obj/no obj values I'm getting when I train tiny-yolo-voc on the standard PASCAL VOC07+12 20-class dataset:
![obj_vs_no-obj](https://user-images.githubusercontent.com/31752998/33040278-975a344c-ce08-11e7-9b8f-cca97fff9da6.png) where blue/yellow are from random weight initialization, and purple/green are from darknet.conv.13 initialization. There is a pretty huge performance difference in the [region] layer performance starting from pre-trained conv weights.

I have seen the obj values going to zero issue before and I don't recall what I did to fix it. If you can't reproduce (IOU of ~61% on darknet detector recall) the tiny-yolo-voc results on the standard 20-class dataset, you may have a build issue.

I've read that some fields are sensitive to spaces, so it's important not to have extraneous spaces. So "classes= 4" may not be OK, even if spaces are OK for floats.

**Comment by EnricoBeltramo** [comment]
I modified voc_label.py in order to have only examples for classs of my interest, I changed line 9 of voc_label.py as:
classes = ["car", "cat", "dog", "person"]
I checked too the cfg file: it doesn't look to have extra spaces.
I the while I tried different combination in order to verify the effects of initialitation. In one of those tests, I extract all the 14 layers from tiny-yolo-voc.weigths, and used as initialitation of neural network. I'm still training, but strangely (because I think only 13 layers shall be used) this seems to give good results:

Region Avg IOU: 0.745535, Class: 0.949261, Obj: 0.346126, No Obj: 0.001994, Avg Recall: 1.000000,  count: 8
Region Avg IOU: 0.536440, Class: 0.982621, Obj: 0.191174, No Obj: 0.001312, Avg Recall: 0.666667,  count: 6
Region Avg IOU: 0.596343, Class: 0.988167, Obj: 0.311734, No Obj: 0.002133, Avg Recall: 0.750000,  count: 12
Region Avg IOU: 0.759364, Class: 0.978084, Obj: 0.216389, No Obj: 0.001508, Avg Recall: 1.000000,  count: 5
Region Avg IOU: 0.589381, Class: 0.990352, Obj: 0.392329, No Obj: 0.001394, Avg Recall: 0.700000,  count: 10
Region Avg IOU: 0.731637, Class: 0.950812, Obj: 0.041588, No Obj: 0.001394, Avg Recall: 1.000000,  count: 8
Region Avg IOU: 0.680600, Class: 0.945364, Obj: 0.274381, No Obj: 0.001342, Avg Recall: 0.800000,  count: 5
Region Avg IOU: 0.691045, Class: 0.821562, Obj: 0.190128, No Obj: 0.001335, Avg Recall: 0.833333,  count: 6
1222: 0.180167, 0.277909 avg, 0.001000 rate, 11.474000 seconds, 78208 images
Loaded: 0.000000 seconds
Region Avg IOU: 0.636168, Class: 0.977733, Obj: 0.356297, No Obj: 0.002776, Avg Recall: 0.857143,  count: 14
Region Avg IOU: 0.627587, Class: 0.994622, Obj: 0.333045, No Obj: 0.001938, Avg Recall: 0.875000,  count: 8
Region Avg IOU: 0.602742, Class: 0.906244, Obj: 0.237939, No Obj: 0.001654, Avg Recall: 0.500000,  count: 6
Region Avg IOU: 0.786130, Class: 0.999218, Obj: 0.334911, No Obj: 0.001348, Avg Recall: 1.000000,  count: 2
Region Avg IOU: 0.729536, Class: 0.953546, Obj: 0.142499, No Obj: 0.001608, Avg Recall: 1.000000,  count: 7
Region Avg IOU: 0.628735, Class: 0.949443, Obj: 0.091498, No Obj: 0.001692, Avg Recall: 0.666667,  count: 15
Region Avg IOU: 0.767259, Class: 0.967316, Obj: 0.297329, No Obj: 0.001697, Avg Recall: 1.000000,  count: 5
Region Avg IOU: 0.669265, Class: 0.994454, Obj: 0.244323, No Obj: 0.002222, Avg Recall: 0.833333,  count: 12
1223: 0.190553, 0.269174 avg, 0.001000 rate, 12.027000 seconds, 78272 images

The training is ongoing, but I already tested with partial results and detection looks to be enough good

**Comment by EnricoBeltramo** [comment]
I stopped at 20000 step, with an avg about about 0.19 and obj around 0.35. I tested with some examples and I had a good detection for involved 4 classes, in some case better than original 20 classes tiny yolo voc.
@shizukachan thank you for you support: it was very helpful to understand better the initialization question.

**Comment by robosina** [comment]
@EnricoBeltramo 
Hi, I am stuck in this problem how to solve your problem?
